store/mockstore/mocktikv/rpc.go

```diff
@@ -388,6 +388,23 @@ func (h *rpcHandler) handleKvRawPut(req *kvrpcpb.RawPutRequest) *kvrpcpb.RawPutR
 	return &kvrpcpb.RawPutResponse{}
 }
 
+func (h *rpcHandler) handleKvRawBatchPut(req *kvrpcpb.RawBatchPutRequest) *kvrpcpb.RawBatchPutResponse {
+	rawKV, ok := h.mvccStore.(RawKV)
+	if !ok {
+		return &kvrpcpb.RawBatchPutResponse{
+			Error: "not implemented",
+		}
+	}
+	keys := make([][]byte, len(req.Pairs))
+	values := make([][]byte, len(req.Pairs))
+	for _, pair := range req.Pairs {
```

Comment:
``` 
for _, pair := range req.Pairs {
		keys = append(keys, pair.Key)
		values = append(values, pair.Value)
}
```
above code looks fine, but this is the hot path. We need to improve its spatial locality and cpu cycle. The better way to do is the following:

```
for int i = 0; i < len(req.Pairs); i++ {
	keys = append(keys,req.Pairs[i].Key)
}
for int i = 0; i < len(req.Pairs); i++ {
	values = append(values, req.Pairs[i].Value)
}
```

Reply:
Do you have any benchmark results?

Reply:
changing from `for _, pair := range req.Pairs {}` to `for int i = 0; i < len(req.Pairs); i++ {}` can be proved by your benchmark which you posted a couple days ago. 

For the second one, it refers to `spatial locality`. The direct answer is no. I do not do any benchmark for it. It is what I learnt from csapp. 

```
for int i = 0; i < len(req.Pairs); i++ {
	keys = append(keys,req.Pairs[i].Key)
}
for int i = 0; i < len(req.Pairs); i++ {
	values = append(values, req.Pairs[i].Value)
}
```
is known as stride-1 reference pattern. 

```
for _, pair := range req.Pairs {
		keys = append(keys, pair.Key)
		values = append(values, pair.Value)
}
```
is known as stride-2 reference pattern. 

As the book(csapp) suggests we need try our best to eliminate stride-N reference pattern and write all code in a fashion of stride-1 pattern.

Reply:
Please do run a benchmark.

Reply:
I don't agree. the `Key` field and `Value` field must be contiguous in memory, since they are in a single struct. However, splitting it into two loops make it jumps further and more times in memory. Also this reduced readability.

Reply:
```
package main

import "testing"

var sLen = 64

type Pair struct {
 key   []byte
 value []byte
}

func BenchmarkStride1(b *testing.B) {
 pairs := make([]*Pair, 0, 1024)
 for i := 0; i < len(pairs); i++ {
  pairs[i] = &Pair{[]byte{byte(i)}, []byte{byte(i)}}
 }

 keys := make([][]byte, sLen)
 values := make([][]byte, sLen)

 b.ResetTimer()
 for i := 0; i < b.N; i++ {
  for j := 0; j < len(pairs); j++ {
   keys[j] = pairs[j].key
  }
  for j := 0; j < len(pairs); j++ {
   values[j] = pairs[j].value
  }
 }
}

func BenchmarkStride2(b *testing.B) {
 pairs := make([]*Pair, 0, 1024)
 for i := 0; i < len(pairs); i++ {
  pairs[i] = &Pair{[]byte{byte(i)}, []byte{byte(i)}}
 }

 keys := make([][]byte, sLen)
 values := make([][]byte, sLen)

 b.ResetTimer()
 for i := 0; i < b.N; i++ {
  for j := 0; j < len(pairs); j++ {
   keys[j] = pairs[j].key
   values[j] = pairs[j].value
  }
 }
} 
```
I just did a benchmark, the result is similar and sometime wrap operation inside the same loop performs better.

Reply:
goos: darwin
goarch: amd64
BenchmarkStride1-8   	2000000000	         1.00 ns/op
BenchmarkStride2-8   	2000000000	         0.66 ns/op
PASS
ok  	command-line-arguments	3.512s

Reply:
No. Original style is actually better than the style I proposed.

Reply:
Another thing I want to say is, I think we should only do this kind of optimization when the performance is really critical.

Reply:
Why so serious, this is only a mocktikv...

Reply:
Why is this code path not cirtical? It seems critical to me. It will be really helpful if you want expalin more.

Reply:
It does not run in a production environment. IMO its readability has a much higher priority than performance.